### PR TITLE
Implement a stable objective UndoRedo layer

### PR DESCRIPTION
**Pull request by ator-dev**
TLDR:
* Normal `create/commit_action` methods introduce crashes, inconsistency and lost time; new optional `create/commit_action_cumulative` methods work automatically to speed up development and enable reliable UndoRedo.
* Cumulative UndoRedo works on what actually happened, not what was meant to happen.

Implement godotengine/godot-proposals#3714 to provide a stable base for godotengine/godot-proposals#2109

On its own this PR does not change engine operation, since it only provides an abstraction layer for the UndoRedo system and doesn't add any use cases. However, my test PR #56770 utilises the new layer. By using it, **many** problems and open issues with Animation editing are fixed by _simplifying_ the code in most places.

This is because I designed the system to 'automatically' work. Contributors using `create_action_cumulative` no longer need to log arbitrary action methods/properties - which creates edge cases, broken functionality and crashes all over the place - instead only needing to add two lines to their code (to create+commit the action). 

For example, in 4.0 and even 3.x, it's currently difficult to do undos/redos in the Animation editor without at some point having keys go missing, get duplicated or break, often resulting in an engine crash; other types of issues are present as well. This is because the methods used don't work the same forwards as backwards, with contributors having to put a lot of effort into finding functions that seem correct and applying them in an order that 'works', with undo and redo being _completely independent_. The smallest change to a method _used by undo or redo_ can introduce inconsistency or crashes, meaning that a lot of development time has to be put into maintaining undo; and yet, as long as Godot is in development it can never work perfectly or even well.

On the other hand, this 'cumulative' UndoRedo system will always revert to the exact same state (as long as `get` and `set` are maintained) requiring next to no contributor effort and eliminating a major cause of crashes/issues slowing down 4.0 development. Even if something unexpected happens - e.g. a malfunction in the editor code breaks something made by the user - UndoRedo can always save the affected object from being lost, because it works on what _actually_ happened, not what was meant to happen.

It's true that in a few places the changes to Animation-editing were larger. However, this is a minority of cases caused by the systems being designed to work with the normal UndoRedo methods, so in some areas a minor redesign was necessary. In some ways this is a benefit though, because I think cumulative-undo encourages better code practices and reusability. In any case, if some area of the engine is too difficult or impractical to use cumulative undo in, it can just continue using the normal system.

### CODE REVIEW

**Comment by ator-dev** [comment]
By the way, please tell me (with a reaction/comment etc) if I should provide some easier way of proving this functionality in order for it to be merged. For example, I could make a side-by-side video for current Animation-editing vs how it works with my changes (or some other comparison video), build the engine with some cumulative undos added and uploading it here, or do something else that shows it's worth adding this to Godot (;

**Comment by groud** [comment]
Sorry but I am not convinced by this approach, it maybe makes writing code faster but it is at the price of significantly slowing down any editor action.

Most of the time this is not a problem anyway, as storing the previous state of a property stays relatively simple. The real complex cases that UndoRedo has to handle are usually when you have modified sub-objects, which your implementation does not covers.
As a consequence, I don't think it is worth making UndoRedo slower considering the most common use case works fines with the current approach, and this new one does not help with more complex scenarios.

**Comment by reduz** [comment]
(I just noticed I commented in the proposal rather than here, so copying it here)

My feelings about this is that, in my experience, most of the time when dealing with undo and redo, you are not really dealing with properties or something that simply changes state. In most cases you are adding or removing things, so its not really possible to do different to the way it is now.

Even in many cases where you modify properties, its not as simple as replacing new value by existing value, as the existing value has to be cached at the beginning of editing (as in when you drag something and move it).

So, even though I understand this can be useful in some cases, it sounds like it would be a minimal amount of them, for which I am not entirely sure it's worth doing this change.

**Comment by ator-dev** [comment]
Thank you very much for your time in checking this and pointing out potential issues! I really appreciate your understanding in dealing with my proposal and considering a differing point of view. Below I have given my responses (and explanations) to the points you brought up - if you have the time please take a moment to look at relevant sections, as I have tried to lay out my argument clearly there in order to better describe how the code applies, and why I believe it genuinely to be a useful addition that could make integration of safe undo systems much simpler. The code screenshots may also be helpful in summing up what I think are 'problem areas' with the current system.

A question I will pose to you first, though, is: Are you happy with the way undo-redo is currently handled, and the code that contributors are making in order to use the system correctly? If so, then I have no objection and you can continue using the main undo methods with no 'cumulative' abstraction/helper layer.

As I see it undo-redo is a chore that in many cases takes a lot of work to make function for a particular part of the editor (e.g. newly designed or redesigned by a contributor), and then can be badly affected by a change in one of the used methods or may lead to failure/discrepancy edge cases; and to me the resulting code looks difficult to understand, maintain, or change. This is only my opinion though - as I would like to emphasise, and I fully acknowledge my bias towards cumulative undo given that I designed it - and you may consider it in a different light; in which case I'd appreciate your thoughts on this. If there's a certain way you'd like contributors to be working with the main undo-redo methods that would work better and isn't always being followed, that would be good to know as well.

@groud 's points:

> at the price of significantly slowing down any editor action

This is a concern I had originally as well, and I can see why you don't like the idea of the system having to check where the changes happened in an object when an action is committed. I would argue though that compared to the price of, for example, rendering text or doing other background processes, this isn't hugely expensive and not something to worry about - all that happens is it checks a list of often less than 10, almost always less than 20 properties and logs any with a difference from the cached value. Personally, there has been no perceivable difference in performance for me; lag from introducing too many objects/keys/values occurs long before and is the same amount as with the normal methods. Also, since this happens once every time the user performs a significant action and has no effect on project performance, it is not too important in my opinion and many of the editor methods already do more complex looping operations, some of which, unlike cumulative undo, might not have linear time complexity.

> storing the previous state of a property stays relatively simple.

I agree completely that the main system works in an intuitive way, and it doesn't take a lot of effort to make a simple action work. However, I think it's very easy to make mistakes - especially where references/pointers are involved - and the resulting code is often hard to make sense of for when it needs to be changed or understood. There are lots of complex if/case statements that deal with undo actions, and the additional complication of needing dedicated variables and/or calls for rollback values can make code significantly less clear and so harder to maintain.

The current system results in strings of nested code, full of do/undo methods often unclear in their relation to each other. As I see it, the problem even greater than unreadable code is that these methods can change internally at any time; there is no restriction on which methods are used, including a mixture of internal processing, abstraction layers, and methods that are exposed to GDScript (or the main APIs).

The methods contain a large amount of their own complexity/processing, some of which could be completely unrelated to the needs of the undo system, and there is no guarantee that 'opposite' methods do the exact opposite thing (especially when there is no 'opposite' method available, so a combination of others must be used). Using the current system there will always be edge cases to fix, where the modified object (or other objects modified by the methods) gain a particular state unexpected by the undo system, leading potentially to a crash or data loss; just one incongruency has the potential to escalate with more undos/redos which the user may automatically try when they notice that something has gone wrong.

Below are some screenshots of code I believe highlights some of the issues particularly well; rollback values can be non-obvious as well as how methods relate, spaghetti code has been generated through having to call long strings of methods, and special cases (often extending a long way above and below) have to be given their own sets of similar but not quite identical undo/redo lines:

image
image
image
image
image
image

> when you have modified sub-objects, which your implementation does not covers.

> does not help with more complex scenarios.

My implementation does not follow object references automatically to avoid unneeded processing; however, making a nested action for a modified sub-object works as expected (as I tested when replacing all undo methods in the animation system). In my opinion this is a good solution, since it makes very clear what objects outside of the main target are being modified, and possibly discourages unnecessary state changes outside of the current object.

You're right that complex scenarios will not necessarily work with this system - however, I think there are [broadly] two main types of 'complex scenario' handled by undo-redo. The first involves creating and moving nodes or similar objects within a data structure like a SceneTree; this would not work very naturally with cumulative undo, though it could probably be made to work. The second is complex because lots of method calls (`do` and `undo`) are made to an object progressing it to a particular desired state, often including calls that only do a very small amount of the processing but which must be accounted for nevertheless. This can be reduced. with cumulative undo`, to only a few lines of code - a state change is declared to have started, the state is changed with a few direct calls, and this is committed with precisely the previous state and the new state being recorded.

If something goes slightly wrong during the main undo methods (relatively likely in this second kind of complex situation) and the state is not returned to the exact expected state, this can leak out with more undos/redos and affect other parts of the object, but cumulative works almost as a backup that can fix the result of a complex operation not quite working (resulting otherwise in data loss or corruption). This could really help with users' trust in the engine and especially undo-redo, which at the moment is not great as demonstrated in [this poll](https://www.reddit.com/r/godot/comments/slwtkx/poll_how_do_you_use_godots_undo_system/?utm_source=share&utm_medium=web2x&context=3); when I have spoken with the community about it in general, they have expressed distrust and unhappiness with the system and it's fairly common to avoid doing more than one undo at a time.

@reduz 's points:

> you are not really dealing with properties or something that simply changes state. In most cases you are adding or removing things

It's true that not many deal with properties explicitly (call `add_do_method` rather than `..._property`), but underneath it is usually simple state that is changed as a result, which cumulative undo works well with. Any places where this system would work especially poorly it doesn't have to be implemented, but since it utilises the methods called when loading scenes/resources from files most things including adding/removing objects works automatically anyway. For example, when I tested replacing all animation-related undo methods with cumulative ones, this included creating and removing animation resources as well as other similar processes, and worked as expected.

> the existing value has to be cached at the beginning of editing

Could you give some more detail on this please? The system works essentially like this, by caching the value at the start of an edit (when `create_action_cumulative` is called) - I tried it with animation key editing [which works by dragging], and instead of breaking the system it fixed several significant issues with it where keys are deleted or moved incorrectly by the main undo system methods.

> even though I understand this can be useful in some cases, it sounds like it would be a minimal amount of them

Thank you for considering my idea and seeing its potential uses! I understand that it looks like it would be non-useful in the majority of situations, but (as mentioned in some of my points above) I have found it to work well in almost every case - the main ones I'm dubious about are the ones involving adding/removing nodes, which is probably best handled by the main undo methods. It functioned with no noticeable performance difference when all animation editor undo-calls were replaced, fixing several problems at the same time, and from looking at many of the other files I see no reason it couldn't be implemented to a large extent in other places as well.

By the way, to demonstrate where cumulative undo could be especially helpful, I tried it on a part of the animation editor which previously had no undo-redo implemented due to how complex it would be to get working: Optimize and Clean Up Anim. It enabled this to be done through adding 2 lines of readable code, and meant that as well as being able to safely cache a complicated action, this action would no longer cause the undo history stack to be deleted which is confusing and frustrating for users.

**Comment by groud** [comment]
> Are you happy with the way undo-redo is currently handled, and the code that contributors are making in order to use the system correctly? Do you think the amount of effort required to maintain actions, and the reliability issues users are having with undo-redo, is acceptable for the editor? 

Well, I never had to complain too much about it. The API is quite simple and straightforward to use. So yeah, I like it the way it is right now. We needed to add a few features here and there for some specific use cases (like, for example, the TileData's object's properties that depend on its owner TileSet's properties). But otherwise it works fine.

> I agree completely that the main system works in an intuitive way, and it doesn't take a lot of effort to make a simple action work. However, I think it's very easy to make mistakes - especially where references/pointers are involved - and the resulting code is often hard to make sense of for when it needs to be changed or understood.

Yeah, and the main problem is that your solution won't solve those problems anyway. The most complex cases are when your editing modifies several objects at the same time, you have visual/cache updates to trigger elsewhere, or optimizations to be done not to slow down the editor, etc... So even with an automatic way to detect which properties changed, then you would still have to do a lot of manual work to solve those complex cases. Your solution only make sense for small objects modifications, and basically transforms a handful of `add_do_*` and `add_undo_*` into a single function call.

For those cases, it might be interesting for some people to write a little bit less code, but in general in Godot, we favor simple APIs in place of more complex ones, even if it means the end user has to write less code. Thus, regarding that point, most of your screenshots show perfectly acceptable and understandable code to me.

**Comment by ator-dev** [comment]
Thank you for confirming this, and thanks again for your time. I personally prefer APIs that allow less code to be written, but I can understand why [in the Godot design philosophy] you prefer more simple ones that give more power and efficiency, and see that this may not be of interest to Godot developers and its design. I'll close the PR with the understanding that either this suggestion or a future one could still supersede the main undo system later on, but at least right now there is little interest in overhauling the current way of working as issues caused by it are acceptable.

Final answers to your points, for the record (don't read if you don't want to / can't spare the time):

> The API is quite simple and straightforward to use.

I agree with this, but not fully, since it does result in contributors having to put a lot of effort into making a complex series of function calls (and reversed operations) work for undo-redo - I see this as difficult to understand and maintain. You're right that it's efficient and easy to understand/control when writing undo-redo code though, and it's true that in many cases it is the 'right' system.

> Your solution only make sense for small objects modifications, and basically transforms a handful of `add_do_*` and `add_undo_*` into a single function call.

Your summary is correct (though unless I misunderstand you're implying that only small changes to objects are covered, whereas I think my system works best for large, complex changes). It does still take some extra work to implement for changes to several objects, several cases (such as adding/removing nodes) aren't covered, and it does only combine do/undo calls. In my opinion this would be helpful for improving maintainability and reliability of certain areas of the editor, but again the worth of this is for project lead developers such as you and reduz to decide, and I accept that Godot's philosophy doesn't necessarily integrate well with this.
